Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,460 @@
+We are in the last function of the chain, so after it does its work, it returns to its CallTarget thunk. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+* Stack arguments of D
+* Return address of D
+* Callee saved registers and locals of D
+```
+### Returned to CallTarget thunk for D with ChainCall flag in the TailCallHelperStackEntry **clear**
+The thunk checks the ChainCall flag and since it is clear, it recognizes we are now returning from the call chain and so it returns the result of the D. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### Returned to A
+We are back in A and we have the return value of the call chain.
+```
+* Return address of A
+* Callee saved registers and locals of A
+```
+
+## Example of thunks generated for a simple generic method
+
+```C#
+struct Point
+{
+    public int x;
+    public int y;
+    public int z;
+}
+
+class Foo
+{
+    public Point Test<T>(int x, T t) where T : class
+    {
+        Console.WriteLine("T: {0}", typeof(t));
+        return new Point(x, x, x);
+    }
+}
+```
+
+For tail calling the Test function, the IL helpers could look e.g. as follows:
+
+```IL
+.method public static void StoreArgumentsTest(native int target, object thisObj, native int genericContext, int x, object t) cil managed
+{
+    .maxstack 4
+    .locals init (
+       [0] native int buffer,
+    )
+
+    call native int AllocateArgumentBuffer(56)
+    stloc.0
+
+    ldloc.0
+    ldc.i8 0x12345678 // pointer to the GC descriptor of the arguments buffer
+    stind.i
+    ldloc.0
+    sizeof native int
+    add
+    stloc.0
+
+    ldloc.0
+    ldftn Point CallTargetTest()
+    stind.i
+    ldloc.0
+    sizeof native int
+    add
+    stloc.0
+
+    ldloc.0
+    ldarg.1 // "thisObj"
+    stobj object
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldarg.2 // "genericContext"
+    stind.i
+    ldloc.0
+    sizeof native int
+    add
+    stloc.0
+
+    ldloc.0
+    ldarg.3 // "x"
+    stind.i4
+    ldloc.0
+    sizeof native int
+    add
+    stloc.0
+
+    ldloc.0
+    ldarg.4 // "t"
+    stobj object
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldarg.0 // "target"
+    stind.i
+
+    ret
+}
+
+```
+
+```IL
+.method public static Point CallTargetTest() cil managed
+{
+    .maxstack 4
+    .locals init (
+       [0] native int buffer,
+       [1] valuetype TailCallHelperStackEntry entry,
+       [2] Point result
+    )
+
+    // Initialize the TailCallHelperStackEntry
+    // chainCall = false
+    // sp = current sp
+
+    ldloca.s 1
+    ldc.i4.0
+    stfld bool TailCallHelperStackEntry::chainCall
+    ldloca.s 1
+    call native int GetCurrentSp()
+    stfld native int TailCallHelperStackEntry::sp
```

Comment:
How do we envision this to be used by the helper and the JIT? Will `GetCurrentSp` be an intrinsic? In my experiment I have implemented the helper with a simple check that `rsp` is equal to this value to figure out whether the current call frame is this one. However `rsp` will be after the `calli` below which is not always trivial to predict (although due to `FEATURE_FIXED_OUT_ARGS` in the JIT it should be for many platforms).

I wonder if it would be easier to use the return address although I don't see exactly how we could do that either.

Reply:
It could be an intrinsic or a JIT helper. It seems the perf difference would be negligible due to the other stuff we need to do to perform the tail call. 
The SP we get in the TailCallHelper should actually not be influenced by the calli parameters even on x86, since on that platform, the callee cleans the stack, so when the TailCallHelper is tail-called by a function that was tail-called too, the arguments should be already removed from the stack and so the SP should match the captured one.

Btw, I have actually planned to use the return address instead in one of the initial iterations of the design. My original plan was that the tail call helper stack entry would store the range of IP addresses of the CallHelper so that I can compare the return address to that range. My guess is that the reason for the switch was that I didn't know if JIT, even with an addition of some intrinsic, can reasonably figure that out. And the SP capturing  / comparison was simpler.

Reply:
> The SP we get in the TailCallHelper should actually not be influenced by the calli parameters even on x86, since on that platform, the callee cleans the stack, so when the TailCallHelper is tail-called by a function that was tail-called too, the arguments should be already removed from the stack and so the SP should match the captured one.

Good point, I have not realized that. 

> Btw, I have actually planned to use the return address instead in one of the initial iterations of the design. My original plan was that the tail call helper stack entry would store the range of IP addresses of the CallHelper so that I can compare the return address to that range. My guess is that the reason for the switch was that I didn't know if JIT, even with an addition of some intrinsic, can reasonably figure that out. And the SP capturing / comparison was simpler.

I still wonder if we can perhaps solve both the problems with generic context and return address using a single intrinsic. Something like
```
ldtoken TargetFunction
ldloc tailCallStackEntry
ldflda TailCallStackEntry::retAddr
ld Args::genericContext
ld Args::retBuffer
ld Args::arg1
ld Args::arg2
...
ld Args::argN
call System.Runtime.JitTailCallDispatchDirect
```
and the JIT would store the return address into `retAddr` and do the necessary setup for the generic context and retbuffer. Initial prototype would probably only have one indirect dispatch intrinsic, but later the direct version (shown above) could be added too allowing for inlining if that was deemed worthwhile.

By the way, I plan to work on implementing (a prototype of) the design in this PR for the rest of my internship.